python/ray/data/dataset.py

```diff
@@ -3985,8 +4008,14 @@ def _get_size_bytes(block: Block) -> int:
 
 
 def _block_to_df(block: Block):
+    from ray.air.util.data_batch_conversion import _cast_tensor_columns_to_ndarrays
+
     block = BlockAccessor.for_block(block)
-    return block.to_pandas()
+    df = block.to_pandas()
+    ctx = DatasetContext.get_current()
+    if ctx.enable_tensor_extension_casting:
+        df = _cast_tensor_columns_to_ndarrays(df)
```

Comment:
Given Pandas block has the tensor casting logic - https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/pandas_block.py#L191-L193 . Do you think we should add the same casting logic for Arrow block as well - https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/arrow_block.py#L191 ? Curious why we have the difference?

Reply:
@c21 Hmm I initially didn't add it to `ArrowBlockAccessor.to_pandas()` since I thought that we used it internally within Datasets and we don't want to do this cast unless the produced DataFrame is user-facing (since it's O(# columns) and makes our processing of the tensor column less efficient)... But I can't find anywhere we use `ArrowBlockAccessor.to_pandas()` other than providing user-facing DataFrames.

Good catch/question, I'll make that change and see if all tests pass as expected.